### ISSUE TITLE
Microsoft SQL Server Engine?

### ISSUE THREAD

**Issue by brianmckeen** [theory/sqitch#149]
Do you have any plans to add an engine for Microsoft SQL Server?

**Comment by drmuey** [comment]
For anyone interested in this issue, there is already a thread (and efforts underway) for this.

https://groups.google.com/forum/#!topic/sqitch-users/RILseUAHRPY

Thanks @theory and @brianmckeen

**Comment by theory** [comment]
@brianmckeen's fork is [here](https://github.com/brianmckeen/sqitch/commits/master).

**Comment by drmuey** [comment]
thanks! @theory

**Comment by theory** [comment]
Hey, what's the status on this, guys?

**Comment by drmuey** [comment]
I've been on a different project (using sqitch but w/ sqlite) but will revisit soon and update this issue, thanks @theory!

**Comment by theory** [comment]
Awesome, I look forward to it, @drmuey.

**Comment by drmuey** [comment]
Quick update for you @theory:

* I just did a pull request on @brianmckeen’s fork to [rename the engine to mssql](https://github.com/brianmckeen/sqitch/pull/4).
* After that I'll start on the driver stuff (I have a [question for you on that at in the group discussion](https://groups.google.com/d/msg/sqitch-users/RILseUAHRPY/3pLfkv2uEgAJ))>

**Comment by theory** [comment]
Awesome, thanks!

**Comment by drmuey** [comment]
Quick update for you @theory:

* @brianmckeen merged the rename-to-mssql pull request, thanks @brianmckeen!
* I just submitted a pull request to @brianmckeen that will [add the mssql DBD driver logic](https://github.com/brianmckeen/sqitch/pull/6).

**Comment by theory** [comment]
Thanks! I left a comment.

**Comment by sudhirvkumar** [comment]
@theory @drmuey Any update on supporting mssql engine?

I see that the latest one seems to be https://github.com/drmuey/sqitch and latest commit is in Jul 2016.

We are using sqitch and would love to be able to use it in our project where we will be using MS SQL Server.

@theory thanks for sqitch, it has made our life better!

**Comment by drmuey** [comment]
@sudhirvkumar The only thing left is adding tests which I stopped working on because of of VPN/VM issues I couldn't resolve (i.e. my tests could not connect to the MSSQL servers I had access to).

The status and punch list at https://github.com/drmuey/sqitch/wiki/MSSQL-TODO

If you're in a position to work out the tests I'd be happy to take a pull request! If we can get it complete then it will be in shape for a proper upstream pull request. I just updated my fork with @theory’s so it should be ready to go!

**Comment by sudhirvkumar** [comment]
@drmuey I will check the code and let you know if I am able to do a PR. Thanks for your reply.

**Comment by jugatsu** [comment]
@sudhirvkumar @drmuey guys, any updates on this issue?

**Comment by GiorgioG** [comment]
+1

**Comment by GiorgioG** [comment]
If there's anything I can do to help, let me know.

**Comment by theory** [comment]
I think someone needs to just pick this back up, rebase it, and get it too completion.

**Comment by drmuey** [comment]
@jugatsu @GiorgioG 

@theory is correct, details of where it is and what needs done are outlined in this comment: https://github.com/theory/sqitch/issues/149#issuecomment-276668457

**Comment by joeandaverde** [comment]
What’s the current state of mssql? 5 years in the making!

**Comment by theory** [comment]
Just takes someone committing to doing the work.

**Comment by sbraverman** [comment]
I have been able to get MsSQL to work using ODBC Driver.
https://github.com/sbraverman/sqitch/tree/mssql-odbc

It builds off all the work that @drmuey has done.

It does not add any tests. 
I am not a Perl-ista (this was my first stab at Perl), nor am I sure that these changes break the way that the ADO driver was using it. 

my sqitch.conf:
```
[core]
	engine = mssql

[engine.mssql]
	driver = {ODBC Driver 17 for SQL Server}
```

The way I run sqitch commands:
```
sqitch verify  -t db:mssql://<user>:<password>@<server>:<port>/<database>
sqitch status  -t db:mssql://<user>:<password>@<server>:<port>/<database>
sqitch deploy  -t db:mssql://<user>:<password>@<server>:<port>/<database>
sqitch revert -t db:mssql://<user>:<password>@<server>:<port>/<database>
```

**Comment by theory** [comment]
NIce. Anyone know of a way to set up automated testing for this on Travis or GH workflow actions?

**Comment by CoreyKaylor** [comment]
You might be able to use containers for this and use https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontainer

https://hub.docker.com/_/microsoft-mssql-server

**Comment by lastlink** [comment]
> You might be able to use containers for this and use https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontainer
> 
> 
> 
> https://hub.docker.com/_/microsoft-mssql-server

I've got several db servers working using service containers see https://github.com/lastlink/dotnet-ci-pipelines/blob/master/.github/workflows/dotnetcore.yml

**Comment by theory** [comment]
Does that docker image include a `sqlcmd` client?

**Comment by lastlink** [comment]
you can use your own docker base image, this is a separate service container docker image that you can access the services of along side the **base** one you are running which you can put whatever you want on.

**Comment by TroySchmidt** [comment]
So the last step is to get the testing in place using containers and the ODBC connection for SQL Server?

**Comment by lastlink** [comment]
I can help with the pipelines, GitHub actions make it easy to run database container with the pipeline.

**Comment by matthewblott** [comment]
MSSQL has been running on Linux for years now, is there a reason it's not supported?

**Comment by lastlink** [comment]
Then again the sql database projects in visual studio work so well (only for sql server) that there probably isn't a need for sqitch to handle it.

**Comment by matthewblott** [comment]
> 

That requires using Windows which I haven't done since about 2015. A lot less people using MS tech work on Windows these days.

**Comment by TroySchmidt** [comment]
> MSSQL has been running on Linux for years now, is there a reason it's not supported?

From above, the work was done to make it work via ODBC but the unit testing for this project was not put in place.  

As for the use case, having one tool to manage database version deployments regardless of engine is nice to have.  Especially for small companies.  While DB projects in Visual Studio does work their dacpac deployment is something of a black box.  It also doesn't have features like being able to tell version or the changelog of applied changes like Sqitch does.

**Comment by arialdomartini** [comment]
Any update on this? Having Sqitch supporting MS SQL would be really great.

**Comment by theory** [comment]
I agree! It just needs someone with the interest, motivation, and time to implement. Do you know of someone who could pick it up?

**Comment by mika76** [comment]
I guess the biggest issue is it's in perl - I'd help if it was in c# or javascript or something 😄

**Comment by theory** [comment]
Well most of the work would need to be in SQL, but I hear ya.

**Comment by mika76** [comment]
@theory I could help with the SQL part if someone else handles the perl part

**Comment by theory** [comment]
I think if you wanted to start with the registry schema, that's be amazing. Should be named `lib/App/Sqitch/Engine/mssql.sql` and be similar to the other SQL files in [that directory](/sqitchers/sqitch/tree/develop/lib/App/Sqitch/Engine). Then maybe the proper arguments to pass to the SQL Server command line client (there is one, yes?) would help to get the basics going.

**Comment by mika76** [comment]
Sure, I'll make a fork and work on that - and yes there's a command line client sqlcmd which can be used. How do you handle different db versions, and how far back do you go? There were some pretty big changes to mssql over the years, although Microsoft only really supports sql 2012+ currently.

**Comment by LucasAlda** [comment]
Any update on this? Having Sqitch supporting MS SQL would be really great.

**Comment by matthewblott** [comment]
@LucasAlda Ha ha, I'd forgotten about this. I just got a notification and I see that I posted here six years ago and this thread is now ten years old! It'd probably make more sense to mark this with the `wontfix` tag.

**Comment by theory** [comment]
I'd still like to see it happen, but someone will have to decide to do it.

**Comment by theory** [comment]
Following up on a [discussion inquiry](https://github.com/sqitchers/sqitch/discussions/607#discussioncomment-13601019), here's a summary of work on SQL Server support up to now (based on a skim through this thread):

* [this fork](https://github.com/brianmckeen/sqitch/commits/master) in 2015
* [this fork(https://github.com/drmuey/sqitch) from 2016
* [this fork](https://github.com/sbraverman/sqitch/tree/mssql-odbc) from 2020 using ODBC ([details](#issuecomment-585486117)

I suspect it could be made ready by a rebase onto `develop` and the introduction of CI testing; I think running an MSSQL Docker image and testing from Linux is a perfectly fine start; borrow from the [Vertica](https://github.com/sqitchers/sqitch/blob/develop/.github/workflows/vertica.yml) and [Exasol]((https://github.com/sqitchers/sqitch/blob/develop/.github/workflows/exasol.yml) workflows.
